# juncmut

The function for analyzing splicing junction associated variant (generated by STAR or bamTojunction, .SJ.out.tab files )

## Dependency

### Require install
 - `pysam`
 - [`junc_utils`](https://github.com/friend1ws/junc_utils)
 - [`annot_utils`](https://github.com/friend1ws/annot_utils)
 - [`edlib`](https://github.com/Martinsos/edlib)

## Install

```
pip install git+https://github.com/ncc-gap/juncmut.git
```

## Preparation of input files

The results from STAR.
format:<br>
chr\<tab>intron_start_position\<tab>intron_end_position\<tab>0\<tab>0\<tab>0\<tab>reads\<tab>reads\<tab>hung_out
1	11845	12009	1	1	0	0	2	47
1	12722	13220	1	1	1	0	3	50

## Preparation of control files.
format:<br>
chr\<tab>start_postion\<tab>end_position(\<tab>any..)

## Requirment of RNA bam

## Commands
### get
```
usage: juncmut get [-h] -input_file input_file -output_file output_file -reference reference -rna_bam rna_bam [-control_file [CONTROL_FILE [CONTROL_FILE ...]]] [-read_num_thres READ_NUM_THRES]
                   [-freq_thres FREQ_THRES] [-mut_num_thres MUT_NUM_THRES] [-mut_freq_thres MUT_FREQ_THRES] -genecode_gene_file genecode_gene_file -output_bam output_bam [-debug debug]

optional arguments:
  -h, --help            show this help message and exit
  -input_file input_file
                        Input file(such as sample.SJ.out.tab file generated by STAR)
  -output_file output_file
                        Output file
  -reference reference  Path to reference genome
  -rna_bam rna_bam      Path to RNA bam file
  -control_file [CONTROL_FILE [CONTROL_FILE ...]]
                        Path to control data created by merge_control (default: None)
  -read_num_thres READ_NUM_THRES
                        Splicing junctions with reads >= read_num_thres is saved (default: 3)
  -freq_thres FREQ_THRES
                        Splicing junctions with reads >= freq_thres is saved (default: 0.05)
  -mut_num_thres MUT_NUM_THRES
                        A mutation with mutation alleles >= mut_num_thres is a true candidate (default: 1)
  -mut_freq_thres MUT_FREQ_THRES
                        A mutation with frequency >= mut_freq_thres is a true candidate (default: 0.05)
  -genecode_gene_file genecode_gene_file
                        Path to genecode gene file
  -output_bam output_bam
                        Output bam
  -debug debug          True keeps the intermediate files.
```
### annot
```
usage: juncmut annot [-h] -input_file input_file -output_file output_file -reference reference -gnomad gnomad [-debug debug]

optional arguments:
  -h, --help            show this help message and exit
  -input_file input_file
                        Input file(such as juncmut.txt generated by juncmut)
  -output_file output_file
                        Output file
  -reference reference  Path to reference genome
  -gnomad gnomad        Path to gnomad vcf file
  -debug debug          True keeps the intermediate files.
```
### validate
```
usage: juncmut validate [-h] -input_file input_file -output_file output_file -reference reference -dna_bam dna_bam [-mut_num_thres MUT_NUM_THRES] [-mut_freq_thres MUT_FREQ_THRES]

optional arguments:
  -h, --help            show this help message and exit
  -input_file input_file
                        Input file obtained by the juncmut get
  -output_file output_file
                        Output file
  -reference reference  Path to reference genome
  -dna_bam dna_bam      Path to DNA bam file
  -mut_num_thres MUT_NUM_THRES
                        A mutation with mutation alleles >= mut_num_thres is a true candidate (default: 1)
  -mut_freq_thres MUT_FREQ_THRES
                        A mutation with frequency >= mut_freq_thres is a true candidate (default: 0.05)
```

## Each function

Step1. Filtering by SJ control and annotation of SJ.  
Step2. Selection the alternative SJ and Adjustment of the position of alternative SJ.              
Step3. Caluculate the frequency and filteration by the frequency.       
Step4. Prediction of mutations.
Step5. Count the number of SJs at a position.  
Step6. Search the mutation in the RNA bam.  
Step7. Realignment  
Step8. Add gnomAD snp annotation.  
-grc option: T means that the chromosome names are not chr-prefix.

## Output col  
Mut_key 
SJ_key 
Sample 
Splicing_class 
SJ_strand 
SJ_read_count 
SJ_depth 
SJ_freq 
Ref_motif: reference 5’ to 3’  
Possive_alt_motif: reference 5’ to 3’ discription. A,T,G,C,R(AG),Y(CT), .(match)  
Possive_alt_key: chr:position,Ref,Alt  
Is_GT_AG 
Is_in_exon 
SJ_overlap_count 
Chr 
Mut_pos 
Mut_ref 
Mut_alt 
Mut_count 
Mut_depth 
Mut_freq 
Realign_no_SJ_neg 
Realign_no_SJ_pos 
Realign_target_SJ_neg 
Reaglin_target_SJ_pos 
Realign_normal_SJ_neg 
Realign_normal_SJ_pos 
Realign_result 
Support_read_rmdup 
RNA_mut 
Gene 
