# juncmut

The function for analyzing splicing junction associated variant (generated by STAR or bamTojunction, .SJ.out.tab files )

## Dependency

### Require install
`pysam`, [`junc_utils`](https://github.com/friend1ws/junc_utils) packages, [`annot_utils`](https://github.com/friend1ws/annot_utils) , `pandas`, sswlib.so


## Install

```
pip install git+https://github.com/ni6o6/juncmut.git
```

Alternatively, you can install from the source code.
```
wget https://github.com/ni6o6/juncmut/dist/juncmut-0.3.1a.tar.gz
tar zxvf .tar.gz
cd
python setup.py install
```

## Preparation of input files

The results from STAR.
format:<br>
chr\<tab>intron_start_position\<tab>intron_end_position\<tab>0\<tab>0\<tab>0\<tab>reads\<tab>reads\<tab>hung_out
1	11845	12009	1	1	0	0	2	47
1	12722	13220	1	1	1	0	3	50

## Preparation of control files.
format:<br>
chr\<tab>start_postion\<tab>end_position(\<tab>any..)

## Requirment of RNA bam

## Commands
### get
```
juncmut get \
-input_file ${input_file} -output_file ${output_file} -reference ${reference} -rna_bam ${rna_bam} \
-control_file ${control_file1} ${control_file2} \
-read_num_thres ${read_num_thres} -freq_thres ${freq_thres} -gnomad ${gnomad} -mut_num_thres ${mut_num_thres} -mut_freq_thres ${mut_freq_thres} 
```
### validate
```
juncmut validate \
-input_file ${input_file} -output_file ${output_file} -dna_bam ${bam} -reference ${reference} -mut_num_thres ${mut_num_thres} -mut_freq_thres ${mut_freq_thres}
-mut_num_thres ${mut_num_thres} -mut_freq_thres ${mut_freq_thres}
```
## docker command  
### get  
```
input_file=/tmp/${sample}.SJ.out.tab
rna_bam=/tmp/${sample}.Aligned.sortedByCoord.out.cram

reference=/tmp/ref/GRCh38.d1.vd1.fa
control_file1=/tmp/ref/SJ_control_2_4_liftHg38_chr.bed.gz
control_file2=/tmp/ref/gtex_hg38_r2s8.bed.gz
read_num_thres=3
freq_thres=0.05
mut_num_thres=2
mut_freq_thres=0.05
gnomad=gs://gnomad-public/release/3.0/vcf/genomes/gnomad.genomes.r3.0.sites.vcf.bgz

docker run -i --rm  -w /tmp -v ~/environment:/tmp \
ni6o6/juncmut:0.2.1 juncmut get \
-input_file ${input_file} -reference ${reference} -rna_bam ${rna_bam} \
-control_file ${control_file1} ${control_file2} \
-read_num_thres ${read_num_thres} -freq_thres ${freq_thres} -mut_num_thres ${mut_num_thres} -mut_freq_thres ${mut_freq_thres} -gnomad ${gnomad}
```
### validate
```
input_file=/tmp/juncmut/${sample}.SJ.fil.annot.assadj.freq.pmut.SJint.rmut.ed.snp.txt
reference=/tmp/ref/GRCh38.d1.vd1.fa
mut_num_thres=2
mut_freq_thres=0.05

docker run -it --rm -w /tmp -v ~/environment:/tmp \
ni6o6/juncmut:0.2.1 juncmut validate \
-input_file ${input_file} -dna_bam ${bam} -reference ${reference} -mut_num_thres ${mut_num_thres} -mut_freq_thres ${mut_freq_thres}
```

## Each function

Step1. Filtering by SJ control and annotation of SJ.  
Step2. Selection the alternative SJ and Adjustment of the position of alternative SJ.              
Step3. Caluculate the frequency and filteration by the frequency.       
Step4. Prediction of mutations.
Step5. Count the number of SJs at a position.  
Step6. Search the mutation in the RNA bam.  
Step7. Realignment  
Step8. Add gnomAD snp annotation.  
-grc option: T means that the chromosome names are not chr-prefix.

## Output col  
Chr  
SJ_Start  
SJ_End  
SJ_Type  
SJ_Strand  
SJ_Read_Count  
SJ_Depth  
SJ_Freq  
Ref_Motif : reference 5’ to 3’  
Possivle_Alt_Motif : reference 5’ to 3’ discription. A,T,G,C,R(AG),Y(CT), .(match)  
Possible_Alt_key : chr:position,Ref,Alt  
Is_GT/AG  
Is_in_exon   
SJ_Overlap_Count  
Mut_Pos  
Mut_Ref  
Mut_Alt  
Mut_Count  
Mut_Depth  
Mut_Freq  
Realign_No_SJ_Neg  
Realign_No_SJ_Pos  
Realign_Target_SJ_Neg  
Reaglin_Target_SJ_Pos  
Realign_Normal_SJ_Neg  
Realign_Normal_SJ_Pos  
RNA_Mut : True or False  
gnomAD  
gnomAD_AF  
g_bases  
g_alt_reads  
g_alt_ratio  
g_mut : True or False  

## bash script
#$ -S /bin/bash
#$ -cwd -V
#$ -l mem_req=16G,s_vmem=16G
#$ -pe def_slot 2


