#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Nov 26 10:30:39 2019
@author: Naoko Iida
"""

from .run_juncmut import run_juncmut
import argparse
#ã€€

def create_parser():
    
    parser = argparse.ArgumentParser(prog = "juncmut")
    
    parser.add_argument("-input_file", metavar = "input_file", default = None, type = str,
                       help = "Input file(sample.SJ.out.tab file generated by STAR)")
    
    parser.add_argument("-reference", metavar = "reference", default = None, type = str,
                        help = "Path to reference genome")
    
    parser.add_argument("-rna_bam", metavar = "rna_bam", default = None, type = str,
                        help = "Path to RNA bam file")
  
    parser.add_argument('-control_file', nargs='*', type = str,
                        help = "Path to control data created by merge_control (default: %(default)s)")
    
    parser.add_argument("-read_num_thres", type = int, default = 3,
                        help = "Remove splicing junctions whose supporting numbers are below this value (default: %(default)s)")
    
    parser.add_argument("-freq_thres", type = float, default = 0.05,
                        help = "Remove splicing junctions whose frequency is below this value (default: %(default)s)")
    
    parser.add_argument("-gnomad", metavar = "gnomad", default = None, type = str,
                        help = "Path to gnomad vcf file")
    
    # set the def    
    parser.set_defaults(func = run_juncmut) 

    return parser
