#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Nov 26 10:30:39 2019

@author: genome
"""

from .run_juncmut import run_juncmut
import argparse
#ã€€

def create_parser():
    
    parser = argparse.ArgumentParser(prog = "juncmut") #make a parser

    # parser.add_argument("input", metavar = "./sampleFolder/sample.SJ.out.tab", default = None, type = str,
    #                     help = "Prefix of input file") 
    
    parser.add_argument("input_SJ", metavar = "sample.SJ.out.tab", default = None, type = str,
                       help = "Path to input splicing junction file (SJ.out.tab file generated by STAR)")

    parser.add_argument("output_file", metavar = "output_file", default = None, type = str,
                        help = "Path to output directory") 

    parser.add_argument("rna_bam", metavar = "rna.bam", default = None, type = str,
                        help = "Path to RNA bam file")

    parser.add_argument("reference", metavar = "reference.fa", default = None, type = str,
                        help = "Path to human reference genome")

    parser.add_argument("gnomad_path", metavar = "gnomad.vcf.gz", default = None, type = str,
                        help = "Path to gnomad vcf file")
    
    parser.add_argument('--control_file', nargs='*', type = str,
                        help = "Path to control data created by merge_control (default: %(default)s)")
    
    parser.add_argument("--read_num_thres", type = int, default = 3,
                        help = "Remove splicing junctions whose supporting numbers are below this value (default: %(default)s)")
    
    parser.add_argument("--freq_thres", type = float, default = 0.05,
                        help = "Remove splicing junctions whose frequency is below this value (default: %(default)s)")
    
    parser.set_defaults(func = run_juncmut) #which def do you use?
    
    
    
    return parser
