# juncmut

[![Build Status](https://travis-ci.org/ni6o6/juncmut/mutation.svg?branch=devel)](https://travis-ci.org/ni6o6/juncmut)

Utility functions for analyzing splicing junction (generated by STAR or bamTojunction, .SJ.out.tab files )

## Dependency

### Require install
`pysam`, [`junc_utils`](https://github.com/friend1ws/junc_utils) packages, [`annot_utils`](https://github.com/friend1ws/annot_utils) , `pandas`,


## Install

```
pip install \<Now not available.>
```

Alternatively, you can install from the source code <Now not available.>
```
wget https://github.com/ni6o6//.tar.gz 
tar zxvf .tar.gz 
cd 
python setup.py install
```

## Preparation of input files
./data/
    |- dir_name/
            |- junction/               <------ The directory to put query files.
            |       |- query files
            |
            |-

filename extension of query files: .SJ.out.tab

format:<br>
chr\<tab>intron_start_position\<tab>intron_end_position\<tab>0\<tab>0\<tab>0\<tab>reads\<tab>reads\<tab>hung_out
1	11845	12009	1	1	0	0	2	47
1	12722	13220	1	1	1	0	3	50

## Preparation of control files.
format:<br>
chr\<tab>start_postion\<tab>end_position(\<tab>any..)

## Preparation of rna bam list.
format:<br>
prefix_of_query_file\<,>path_to_bam

## Commands
```
juncmut prefix_of_query_file dir_name --control_file path_to_control1.bed.gz path_to_control2.bed.gz  --genome_id hg19/hg38 --read_num_thres N --freq_thres N --rbam_chr_prefix none/chr --rbam rna_bam_list.txt
```
prefix_of_query_file (prefix.SJ.out.tab) 
dir_name (./data/dir_name/junction/query files)

### Each function
```
juncmut_env.py {dir_name}
```
```
juncmut_juncutils.py {prefix_of_query_file} {dir_name} --control_file ./reference/SJ_control_2_4.bed.gz ./reference/gtex_sj2r2s.bed.gz
```
            output(./alterativeSJ_fil_annot/prefix.SJ.fil.txt, prefix.SJ.fil.annot.txt )
                 
```
juncmut_assadj.py  {prefix_of_query_file} {dir_name}
```
            output(./alterativeSJ_assadjfreq/prefix.SJ.fil.annot.assadj.txt
            ,prefix.SJ.fil.annot.assadj.txt
            
```
juncmut_freq.py  {prefix_of_query_file} {dir_name} --read_num_thres 3 --freq_thres 0.05
```
           output(./alterativeSJ_assadjfreq/prefix.SJ.fil.annot.assadjunifreq.txt
           ,prefix.SJ.fil.annot.assadjunifreqT.txt

```
juncmut_mutpre.py {prefix_of_query_file} {dir_name} --genome_id hg19
```
            output(./alterativeSJ_mutprediction/prefix.SJ.fil.annot.assadjunifreqT.pmut.txt)

```
juncmut_intersect.py {prefix_of_query_file} {dir_name}
```
            output(./alterativeSJ_mutprediction/prefix.SJ.fil.annot.assadjunifreqT.pmut.SJinSJ.txt)
```
juncmut_annotgnomadsnp.py {prefix_of_query_file} {dir_name} --genome_id hg19
```
            output(./alterativeSJ_mutprediction/prefix.SJ.fil.annot.assadjunifreqT.pmut.SJinSJ.snp.txt)
```
juncmut_annotrnamut.py {prefix_of_query_file} {dir_name} --genome_id hg19 --rbam rna_bam_list.txt
```
            output(./alterativeSJ_mutprediction/prefix.SJ.fil.annot.assadjunifreqT.pmut.SJinSJ.snp.rmut.txt)

optional:<br>
juncmut_annotgmut.py {prefix_of_query_file} {dir_name} --genome_id hg19 --gbam g_bam_list.txt


