# juncmut

Utility functions for analyzing splicing junction (generated by STAR or bamTojunction, .SJ.out.tab files )

## Dependency

### Require install
`pysam`, [`junc_utils`](https://github.com/friend1ws/junc_utils) packages, [`annot_utils`](https://github.com/friend1ws/annot_utils) , `pandas`, sswlib.so


## Install

```
pip install \<Now not available.>
```

Alternatively, you can install from the source code <Now not available.>
```
wget https://github.com/ni6o6//.tar.gz
tar zxvf .tar.gz
cd
python setup.py install
```

## Preparation of input files


|- dir_name/<br>
　　　　|- junction/    <------ The directory to put query files.<br>
　　　　|- juncmut  <------ results

filename extension of query files: .SJ.out.tab

format:<br>
chr\<tab>intron_start_position\<tab>intron_end_position\<tab>0\<tab>0\<tab>0\<tab>reads\<tab>reads\<tab>hung_out
1	11845	12009	1	1	0	0	2	47
1	12722	13220	1	1	1	0	3	50

## Preparation of control files.
format:<br>
chr\<tab>start_postion\<tab>end_position(\<tab>any..)

## Preparation of rna bam list.
format:<br>
prefix_of_query_file\<,>path_to_bam

## Commands

```
juncmut prefix_of_query_file dir_name --control_file path_to_control1.bed.gz path_to_control2.bed.gz  --genome_id hg19/hg38 --read_num_thres N --freq_thres N --rbam_chr_prefix none/chr --rbam rna_bam_list.txt
```
prefix_of_query_file (prefix.SJ.out.tab)
dir_name (./data/dir_name/junction/query files)

### Each function
Step1. You are in the directory with a junction folder.
```
juncmut_env.py {dir_name}
```

Step2. filtering by SJ control and annotation of SJ.

```
juncmut_juncutils.py {input_file} {output_file} --control_file ./reference/SJ_control_2_4.bed.gz ./reference/gtex_sj2r2s.bed.gz
```
output(./juncmut/prefix.SJ.fil.txt, ./juncmut/prefix.SJ.fil.annot.txt )

Step3. Adjustment of the position of SJ.              
```
juncmut_assadj.py  {input_file} {output_file}
```
output(./juncmut/prefix.SJ.fil.annot.assadj.txt)


Step4. Caluculate the frequency and filteration by the frequency.       
```
juncmut_freq.py  {input_file} {output_file} {original_sj_file} --read_num_thres 3 --freq_thres 0.05
```
output(./juncmut/prefix.SJ.fil.annot.assadj.freq.txt)

Step5. Prediction of mutations.
```
juncmut_mutpre.py {input_file} {output_file} {reference}
```
output(./juncmut/prefix.SJ.fil.annot.assadj.freq.pmut.txt)

Step6. Count the number of SJs at a position.
```
juncmut_intersect.py {input_file} {output_file} {original_sj_file}
```
output(./juncmut/prefix.SJ.fil.annot.assadj.freq.pmut.SJinSJ.txt)

Step7. Search the mutation in the RNA bam.
```
juncmut_rnamut.py {input_file} {output_file} {rna_bam} {reference}
```
output(./juncmut/prefix.SJ.fil.annot.assadj.freq.pmut.SJinSJ.snp.rmut.txt)  
column =
[0]'CHR'
[1]'start'
[2]'end'
[3]'start_ori'
[4]'end_ori'
[5]'sample'
[6]'class'
[7]'strand'
[8]'reads'
[9]'total'
[10]'freq'
[11]'ref_bases'
[12]'mut_prediction_seq'
[13]'info'
[14]'type'
[15]'intSJ'
[16]'POS'
[17]'REF'
[18]'MUT'
[19]rna_bases
[20]rna_alt_reads
[21]rna_alt_ratio
[22]rna_mut

Step10. Realign
python ${root}juncmut_realign.py ${sample}.SJ.fil.annot.assadj.freq.pmut.SJinSJ.rmut.txt ${sample}.SJ.fil.annot.assadj.freq.pmut.SJinSJ.rmut.val.txt rna_bam ../../reference/GRCh38_chr.fa hg38 F

-grc option: T means that the chromosome names are not chr-prefix.

output col
[0] Chr	
[1] SJ_Start
[2] SJ_End	
[3] SJ_Type	
[4] SJ_Strand	
[5] SJ_Read_Count	
[6] SJ_Depth	
[7] SJ_Freq	
[8] Ref_Motif	
[9] Possivle_Alt_Motif
[10] Possible_Alt_key	
[11] Is_Canonical	
[12] SJ_Overlap_Count	
[13] Mut_Pos	
[14] Mut_Ref	
[15] Mut_Alt	
[16] Mut_Count	
[17] Mut_Depth	
[18] Mut_Freq	
[19] Realign_No_SJ_Neg	
[20] Realign_No_SJ_Pos	
[21] Realign_Target_SJ_Neg	
[22] Reaglin_Target_SJ_Pos	
[23] Realign_Normal_SJ_Neg	
[24] Realign_Normal_SJ_Pos

Step9. Add gnomAD snp annotation.
```
juncmut_annotgnomad.py {input_file} {output_file} {path_to_gnomad_db} {genome_id} --genome_id hg38
```
output(./juncmut/prefix..SJ.fil.annot.assadj.freq.pmut.SJint.rmut.snp.txt)
column = ['CHR','start','end','start_ori','end_ori','sample','class','strand','reads', 'total', 'freq', 'ref_bases','mut_prediction_seq','info', 'type', 'intSJ','POS','REF','MUT',rna_bases,rna_alt_reads,rna_alt_ratio,rna_mut, gnomAD, gnomAD_AF]

python ${root}juncmut_annotspliceai.py --spliceai /homeB/naiida/project2/reference/spliceai_scores.hg38.vcf.gz --input_file ./juncmut/${sample}.SJ.fil.annot.assadj.freq.pmut.SJint.rmut.val.snp.txt --output_file ./juncmut/${sample}.SJ.fil.annot.assadj.freq.pmut.SJint.rmut.val.snp.spliceai.txt

Optional:  
Step1
```
juncmut_annotgmut.py {input_file} {output_file} {bam} {reference} {rmchr}
```
python ${root}juncmut_annotgmut.py ./juncmut/${sample}.SJ.fil.annot.assadj.freq.pmut.SJint.rmut.val.snp.txt ./juncmut/${sample}.SJ.fil.annot.assadj.freq.pmut.SJint.rmut.val.snp.gmut.txt ${bam} ${reference} T

Step2. Summarize the existance of mutations for alternative SJs.
  filter by The number of intersected SJ
```
agg_1file.py
```




## bash script
#$ -S /bin/bash
#$ -cwd -V
#$ -l mem_req=16G,s_vmem=16G
#$ -pe def_slot 2

root=../../juncmut_codes/
sample=$1
rna_bam=$2
#rna_bam=/homeB/naiida/project2/rna/lung26GRCh38pass1/DRR016694.Aligned.sortedByCoord.out.bam

date

python ${root}juncmut_env.py lung1GRCh38

python ${root}juncmut_juncutils.py ./junction/${sample}.SJ.out.tab ./juncmut/${sample}.SJ.fil.annot.txt --control_file ../../reference/SJ_control_2_4_liftHg38_chr.bed.gz ../../reference/gtex_sj2r2s_liftHg38_chr.bed.gz --genome_id hg38

python ${root}juncmut_assadj.py ./juncmut/${sample}.SJ.fil.annot.txt ./juncmut/${sample}.SJ.fil.annot.assadj.txt

python ${root}juncmut_freq.py ./juncmut/${sample}.SJ.fil.annot.assadj.txt ${sample}.SJ.fil.annot.assadj.freq.txt ./junction/${sample}.SJ.out.tab --read_num_thres 3 --freq_thres 0.05

python ${root}juncmut_mutpre.py ./juncmut/${sample}.SJ.fil.annot.assadj.freq.txt ./juncmut/${sample}.SJ.fil.annot.assadj.freq.pmut.txt ./reference/GRCh38_chr.fa

python ${root}juncmut_intersect.py ./juncmut/${sample}.SJ.fil.annot.assadj.freq.pmut.txt ./juncmut/${sample}.SJ.fil.annot.assadj.freq.pmut.SJint.txt ./junction/${sample}.SJ.out.tab

python ${root}juncmut_rnamut.py ./juncmut/${sample}.SJ.fil.annot.assadj.freq.pmut.SJint.txt ./juncmut/${sample}.SJ.fil.annot.assadj.freq.pmut.SJint.rmut.txt ${rna_bam} ../../reference/GRCh38_chr.fa

python ${root}juncmut_realign.py ./juncmut/${sample}.SJ.fil.annot.assadj.freq.pmut.SJint.rmut.txt ./juncmut/${sample}.SJ.fil.annot.assadj.freq.pmut.SJint.rmut.val.txt ${rna_bam} ../../reference/GRCh38_chr.fa hg38 F

python ${root}juncmut_annotgnomad.py ./juncmut/${sample}.SJ.fil.annot.assadj.freq.pmut.SJint.rmut.val.txt ./juncmut/${sample}.SJ.fil.annot.assadj.freq.pmut.SJint.rmut.val.snp001.txt ../../reference/gnomad.genomes.r3.0.sites.vcf.bgz --genome_id hg38

python ${root}juncmut_annotspliceai.py --spliceai /homeB/naiida/project2/reference/spliceai_scores.hg38.vcf.gz --input_file ./juncmut/${sample}.SJ.fil.annot.assadj.freq.pmut.SJint.rmut.val.snp.txt --output_file ./juncmut/${sample}.SJ.fil.annot.assadj.freq.pmut.SJint.rmut.val.snp.spliceai.txt



date
